ide/lsp.client/src/org/netbeans/modules/lsp/client/bindings/Formatter.java

```diff
@@ -0,0 +1,153 @@
+        Object stream = ctx.document().getProperty(Document.StreamDescriptionProperty);
+        if(stream instanceof MultiDataObject) {
+            MultiDataObject mdo = ((MultiDataObject) stream);
+            FileObject fo = mdo.getPrimaryFile();
+            LSPBindings bindings = LSPBindings.getBindings(fo);
+            if (bindings != null) {
+                Boolean documentFormatting = bindings.getInitResult().getCapabilities().getDocumentFormattingProvider();
+                Boolean rangeFormatting = bindings.getInitResult().getCapabilities().getDocumentRangeFormattingProvider();
+                if (rangeFormatting != null && rangeFormatting) {
+                    rangeFormat(fo, bindings);
+                } else if (documentFormatting != null && documentFormatting) {
+                    documentFormat(fo, bindings);
+                }
+            }
+        }
+    }
+
+    private void rangeFormat(FileObject fo, LSPBindings bindings) throws BadLocationException {
+        DocumentRangeFormattingParams drfp = new DocumentRangeFormattingParams();
+        drfp.setTextDocument(new TextDocumentIdentifier(Utils.toURI(fo)));
+        drfp.setOptions(new FormattingOptions(
+            IndentUtils.indentLevelSize(ctx.document()),
+            IndentUtils.isExpandTabs(ctx.document())));
+        drfp.setRange(new Range(
+            Utils.createPosition(ctx.document(), ctx.startOffset()),
+            Utils.createPosition(ctx.document(), ctx.endOffset())));
+        List<? extends TextEdit> edits = Collections.EMPTY_LIST;
+        try {
+            edits = new ArrayList<>(bindings.getTextDocumentService().rangeFormatting(drfp).get());
+        } catch (InterruptedException | ExecutionException ex) {
+            LOG.log(Level.INFO,
+                String.format("LSP document rangeFormat failed for {0}", fo),
+                ex);
+        }
+        Collections.sort(edits, rangeReverseSort);
+
+        applyTextEdits(edits);
+    }
+
+    private void documentFormat(FileObject fo, LSPBindings bindings) throws BadLocationException {
+        DocumentFormattingParams dfp = new DocumentFormattingParams();
+        dfp.setTextDocument(new TextDocumentIdentifier(Utils.toURI(fo)));
+        dfp.setOptions(new FormattingOptions(
+            IndentUtils.indentLevelSize(ctx.document()),
+            IndentUtils.isExpandTabs(ctx.document())));
+        List<? extends TextEdit> edits = Collections.EMPTY_LIST;
+        try {
+            edits = new ArrayList<>(bindings.getTextDocumentService().formatting(dfp).get());
+        } catch (InterruptedException | ExecutionException ex) {
+            LOG.log(Level.INFO,
+                String.format("LSP document format failed for {0}", fo),
+                ex);
+        }
+        applyTextEdits(edits);
+    }
+
+    private void applyTextEdits(List<? extends TextEdit> edits) throws BadLocationException {
+        Collections.sort(edits, rangeReverseSort);
```

Comment:
Would it make sense to call Utils.applyEditsNoLock here? Could be a new overload taking "start" and "end" offsets; or a predicate like "BiPredicate<Integer, Integer> acceptor", which would then be used to filter out the interesting edits (although the latter feels fairly ugly). Would help in case we find a bug in the way the edits are applied.

(the rangeReverseSort is a nicer variant of what is in Utils.applyEditsNoLock, so could be moved into Utils and be used instead of the current code.)